# Classification of Balanced Scale data set 
the main aim of this project is to classify the balanced scale based on the  [balanced_dataset](http://archive.ics.uci.edu/ml/datasets/balance+scale) which is present in the [UC repository](http://archive.ics.uci.edu/ml/index.php)



we will use different kind of tree classifiers.

## DataSet information
This data set was generated to model psychological experimental results. Each example is classified as having the balance scale tip to the right, tip to the left, or be balanced. The attributes are the left weight, the left distance, the right weight, and the right distance. The correct way to find the class is the greater of (left-distance * left-weight) and (right-distance * right-weight). If they are equal, it is balanced.


## Attribute Information:

1. Class Name: 3 (L, B, R) 
2. Left-Weight: 5 (1, 2, 3, 4, 5) 
3. Left-Distance: 5 (1, 2, 3, 4, 5) 
4. Right-Weight: 5 (1, 2, 3, 4, 5) 
5. Right-Distance: 5 (1, 2, 3, 4, 5)

In [42]:
from IPython.display import HTML

HTML('''<script>
code_show=true; 
function code_toggle() {
 if (code_show){
 $('div.input').hide();
 } else {
 $('div.input').show();
 }
 code_show = !code_show
} 
$( document ).ready(code_toggle);
</script>
<form action="javascript:code_toggle()"><input type="submit" value="Click here to toggle on/off the raw code."></form>''')

### Libraries

In [6]:
# importing of libraries
import pandas as pd

### Reading the dataset

In [40]:
# reading of csv file
df=pd.read_csv("https://archive.ics.uci.edu/ml/machine-learning-databases/balance-scale/balance-scale.data",names=['label','left_weight','left_dist','right_weight','right_dist'])

In [41]:
# show the dataset using .head
print("Dataset")
df.head()

Dataset


label  left_weight  left_dist  right_weight  right_dist
0     B            1          1             1           1
1     R            1          1             1           2
2     R            1          1             1           3
3     R            1          1             1           4
4     R            1          1             1           5

In [14]:
print "shape of dataset"
df.shape

shape of dataset


(625, 5)

In [15]:
print  "size of dataset"
df.size

size of dataset


3125

In [16]:
print "Description of Dataframe"
df.describe(include='all')

Description of Dataframe


label   left_dist  left_weight  right_dist  right_weight
count    625  625.000000   625.000000  625.000000    625.000000
unique     3         NaN          NaN         NaN           NaN
top        R         NaN          NaN         NaN           NaN
freq     288         NaN          NaN         NaN           NaN
mean     NaN    3.000000     3.000000    3.000000      3.000000
std      NaN    1.415346     1.415346    1.415346      1.415346
min      NaN    1.000000     1.000000    1.000000      1.000000
25%      NaN    2.000000     2.000000    2.000000      2.000000
50%      NaN    3.000000     3.000000    3.000000      3.000000
75%      NaN    4.000000     4.000000    4.000000      4.000000
max      NaN    5.000000     5.000000    5.000000      5.000000

In [17]:
print "checking of null values in the dataframe"
df.isnull().sum()

checking of null values in the dataframe


label           0
left_dist       0
left_weight     0
right_dist      0
right_weight    0
dtype: int64

In [18]:
print " store the features of dataset in X "
X=df.iloc[:,1:]

 store the features of dataset in X 


In [19]:
print " store the label or target of dataset in y "
y=df.iloc[:,0]

 store the label or target of dataset in y 


### train_test_split
This is available in the sklearn library and is used for splitting the target and features into train(features,target) and test(features,target).

In [20]:
## To Split the dataset into training and testing sets import train_test_split
from sklearn.cross_validation import train_test_split

C:\Users\Dell\Anaconda2\lib\site-packages\sklearn\cross_validation.py:41: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


In [21]:
# breaking dataset as train and test set
x_train,x_test,y_train,y_test=train_test_split(X,y,test_size=0.40,random_state=42)

## DecisionTreeClassifier

In [22]:
from sklearn.tree import DecisionTreeClassifier

### Using 'gini index' as criterion

In [23]:
clf=DecisionTreeClassifier(criterion='gini')

In [24]:
clf.fit(x_train,y_train)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

In [25]:
pred=clf.predict(x_test)

#### Accuracy  by using gini index

In [26]:
print clf.score(x_test,y_test)

0.772


### using 'entropy' as criterion

In [28]:
clf2=DecisionTreeClassifier(criterion='entropy')

In [30]:
clf2.fit(x_train,y_train)

DecisionTreeClassifier(class_weight=None, criterion='entropy', max_depth=None,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=None,
            splitter='best')

#### Accuracy using 'entropy' criterion

In [31]:
clf2.score(x_test,y_test)

0.77600000000000002

## RandomForestClassifier under ensemble models

In [27]:
from sklearn.ensemble import RandomForestClassifier

### using 'gini index' as criterion

In [28]:
ran=RandomForestClassifier(criterion='gini')

In [29]:
ran.fit(x_train,y_train)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='gini',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=1,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False)

#### accuracy by gini index

In [30]:
ran.score(x_test,y_test)

0.80800000000000005

## using 'entropy' criterion

In [36]:
ran2=RandomForestClassifier(criterion='entropy')

In [37]:
ran2.fit(x_train,y_train)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='entropy',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=1,
            oob_score=False, random_state=None, verbose=0,
            warm_start=False)

### accuracy by 'entropy'

In [38]:
ran2.score(x_test,y_test)

0.77200000000000002

## ExtraTreesClassifier under ensemble models

In [33]:
from sklearn.ensemble import ExtraTreesClassifier

## using 'gini index' as criterion

In [34]:
ext=ExtraTreesClassifier(criterion='gini')

In [35]:
ext.fit(x_train,y_train)

ExtraTreesClassifier(bootstrap=False, class_weight=None, criterion='gini',
           max_depth=None, max_features='auto', max_leaf_nodes=None,
           min_impurity_decrease=0.0, min_impurity_split=None,
           min_samples_leaf=1, min_samples_split=2,
           min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=1,
           oob_score=False, random_state=None, verbose=0, warm_start=False)

### Accuracy by gini index

In [36]:
ext.score(x_test,y_test)

0.80400000000000005

## using 'entropy' as criterion

In [37]:
ext=ExtraTreesClassifier(criterion='entropy')

In [38]:
ext.fit(x_train,y_train)

ExtraTreesClassifier(bootstrap=False, class_weight=None, criterion='entropy',
           max_depth=None, max_features='auto', max_leaf_nodes=None,
           min_impurity_decrease=0.0, min_impurity_split=None,
           min_samples_leaf=1, min_samples_split=2,
           min_weight_fraction_leaf=0.0, n_estimators=10, n_jobs=1,
           oob_score=False, random_state=None, verbose=0, warm_start=False)

### Accuracy by entropy

In [39]:
ext.score(x_test,y_test)

0.78400000000000003

## Summary
As by using different types of classifiers like DecisionTreeClassifier,RandomForestClassifier,'ExtraTreesClassifier' we find the acuracy at two different criterion 'gini index' and 'entropy' and we observe that in both criterion 'ExtraTreesClassifier' gives better accuracy as 0.80400(gini index) and 0.78400(entropy) .